versioned/persist/store/src/main/java/org/projectnessie/versioned/persist/store/PersistVersionStore.java

```diff
@@ -0,0 +1,265 @@
+                      storeWorker.getType(op.getExpectedValue())));
+            }
+            if (!contentsId.equals(ContentsId.of(storeWorker.getId(op.getExpectedValue())))) {
+              throw new IllegalArgumentException(
+                  String.format(
+                      "Conditional put-operation key '%s' has different contents-ids.",
+                      op.getKey()));
+            }
+
+            expectedValue = Optional.of(storeWorker.toStoreGlobalState(op.getExpectedValue()));
+          } else {
+            expectedValue = Optional.empty();
+          }
+          commitAttempt.putExpectedStates(contentsId, expectedValue);
+          commitAttempt.putGlobal(contentsId, newState);
+        } else {
+          if (op.getExpectedValue() != null) {
+            // TODO shall we throw this exception? shall we validate the value somehow?
+            //  or just document the behavior?
+            throw new IllegalArgumentException(
+                String.format(
+                    "Contents-type '%s' for put-operation for key '%s' does not support global state, expected-value not supported for this contents-type.",
+                    storeWorker.getType(op.getValue()), op.getKey()));
+          }
+        }
+      } else if (operation instanceof Delete) {
+        commitAttempt.addDeletes(operation.getKey());
+      } else if (operation instanceof Unchanged) {
+        commitAttempt.addUnchanged(operation.getKey());
+      } else {
+        throw new IllegalArgumentException(String.format("Unknown operation type '%s'", operation));
+      }
+    }
+
+    commitAttempt.commitMetaSerialized(storeWorker.getMetadataSerializer().toBytes(metadata));
+
+    return databaseAdapter.commit(commitAttempt.build());
+  }
+
+  @Override
+  public void transplant(
+      BranchName targetBranch, Optional<Hash> referenceHash, List<Hash> sequenceToTransplant)
+      throws ReferenceNotFoundException, ReferenceConflictException {
+    databaseAdapter.transplant(targetBranch, referenceHash, sequenceToTransplant);
+  }
+
+  @Override
+  public void merge(Hash fromHash, BranchName toBranch, Optional<Hash> expectedHash)
+      throws ReferenceNotFoundException, ReferenceConflictException {
+    databaseAdapter.merge(fromHash, toBranch, expectedHash);
+  }
+
+  @Override
+  public void assign(NamedRef ref, Optional<Hash> expectedHash, Hash targetHash)
+      throws ReferenceNotFoundException, ReferenceConflictException {
+    databaseAdapter.assign(ref, expectedHash, targetHash);
+  }
+
+  @Override
+  public Hash create(NamedRef ref, Optional<Hash> targetHash)
+      throws ReferenceNotFoundException, ReferenceAlreadyExistsException {
+    if (ref instanceof TagName && !targetHash.isPresent()) {
+      throw new IllegalArgumentException(
+          "Tag-creation requires a target named-reference and hash.");
+    }
+
+    return databaseAdapter.create(ref, targetHash.orElseGet(databaseAdapter::noAncestorHash));
```

Comment:
Just wondering: why we default the branch to `noAncestorHash` but deny tag creation without an explicit `noAncestorHash`? To me branches and tags are very similar (exception that one cannot commit to tags). I'd think default ancestor behaviour ought to be the same for them (either deny in both cases, or default to `noAncestorHash` in both cases).

Reply:
If this is a usability-level check, perhaps we could move it to the REST API so as to make it effective across the board?

Reply:
While you can create a branch that points to the "beginning of time" (aka that `noAncestorHash`, that does not exist in the database), you cannot do that for a tag. (Not a new behavior though)

Moved that check to `TreeResource`